### PR TITLE
[Dashboard] Name the managed-jobs filter params in the URL

### PR DESCRIPTION
**Pull request by zpoint**
> Builds on the hook and schema from #10455, now merged — this PR is one file (`jobs.jsx`).

## Problem

The jobs page kept its most-used filter entirely out of the URL. Clicking
**Active**, then **RUNNING**, then **SUCCEEDED** produces three different views
and leaves the address bar at `/dashboard/jobs` throughout — so "hey, look at all
the failed jobs here: …" is a link that shows the recipient the default page.

The chip filters were in the URL, but as positional triples
(`?property=user&operator=%3A&value=alice`).

## What a link looks like

Driven through the UI on a local server with twelve seeded managed jobs across
six statuses and three users. **Row counts are identical at every step — only the
URL changes.**

| What the user does | Before | After | Rows |
|---|---|---|---|
| Open the page | `/dashboard/jobs` | `?owner=all` | 10 |
| Click **Active** | `/dashboard/jobs` — unchanged | `?owner=all&status=active` | 6 |
| Click the **RUNNING** pill | `/dashboard/jobs` — unchanged | `?owner=all&status=RUNNING` | 4 |
| Also click **SUCCEEDED** | `/dashboard/jobs` — unchanged | `?owner=all&status=RUNNING,SUCCEEDED` | 6 |
| Filter `User = alice` | `?property=user&operator=%3A&value=alice` | `?user=alice` | — |
| Open `?owner=all&status=FAILED,CANCELLED` | not expressible | opens on exactly those jobs | 4 |

That last row is the ticket's original ask, and it is the one that could not be
built at all before.

### Before

1. initial: /dashboard/jobs, 6 rows  `http://100.86.230.128:46621/dashboard/jobs`

image

2. Active: 3 rows, address bar STILL empty `http://100.86.230.128:46621/dashboard/jobs`

image

3. RUNNING pill: 2 rows, address bar STILL empty `http://100.86.230.128:46621/dashboard/jobs`

image

4. + SUCCEEDED pill: 3 rows, address bar STILL empty `http://100.86.230.128:46621/dashboard/jobs`

image

### After

1. initial: /dashboard/jobs, 10 rows  `http://100.86.230.128:46623/dashboard/jobs?owner=all`

image

2. Active: 3 rows `http://100.86.230.128:46623/dashboard/jobs?owner=all&status=active`

image

3. RUNNING pill: 2 rows, address bar STILL empty `http://100.86.230.128:46623/dashboard/jobs?owner=all&status=RUNNING`

image

4. + SUCCEEDED pill: 3 rows, `http://100.86.230.128:46623/dashboard/jobs?owner=all&status=RUNNING,SUCCEEDED`

image

## Change

`status` becomes the single source of truth for which statuses the table shows;
the pill bar and the Active/All segments are two views of it:

```
absent               every status
active / finished    that group, named rather than expanded, so the link stays
                     short and keeps meaning "the active ones" rather than
                     freezing today's membership of the group
FAILED,CANCELLED     exactly those, which is what clicking pills builds
```

This retires `showAllMode`, which existed only to mark "a pill narrowed the view,
so highlight neither segment" — now derivable from `status` itself. Its other
documented state ("show no jobs") was unreachable: deselecting the last pill
always reset it.

`owner` (My Jobs / All Jobs) joins the same scheme, and clicking a user in the
table replaces rather than stacks a `User` chip, so the page cannot show more
filters than a link is able to carry.

One deliberate behaviour change falls out of this. Deselecting the last pill
clears `status`, which means every status. On master it restored `activeTab`,
which defaults to `'all'` — so the common flow (arrive, click RUNNING, click it
again) is unchanged, but a user who first clicked **Active** now widens to every
status instead of returning to Active. Recovering that needs the group and the
pills to be separate pieces of state, which is what the URL is replacing; and it
is not silent — the **All** segment lights up as the finished jobs appear.

A `status` value the page cannot interpret is ignored rather than trusted, the
same way the filter decoders already drop unknown properties: unknown status
names are filtered out, and the tab falls back to **All** whenever no pill
narrows the view. Without that, `?status=,` rendered an error page and
`?status=RUNNINGG` drew a fake zero-count status chip.

Sort, page and page size stay as they are, matching #10455: their state lives in
the table component and wants its own change.

## Verification

Statuses reach the server through the existing `statuses` parameter, which this
PR does not change: it only changes how the same list is spelled in the address
bar. Checked on a running server that each link in the table above returns the
rows it names, including `?owner=all&status=FAILED,CANCELLED` arriving cold in a
fresh tab.

Two new suites cover the paths that had none, and both fail against the previous
code:

- `shared/FilterDropdown.test.jsx` drives chip add / remove / clear through the
  real components. These widgets take the URL writer as an optional prop, and
  calling it unconditionally crashed every page that does not pass one.
- `jobs.statusview.test.jsx` covers the `status` values a hand-edited or stale
  link can carry, and asserts the invariant that the activity tab is either
  `all` or a key that exists in `statusGroups`.

`npx jest` → 11 suites pass. `version-display` and `jobs.test.jsx` fail identically
on master without this change; the latter is a module cycle between
`data/connectors/jobs` and `plugins/dataEnhancement`, which is also why the new
jobs test has to cut that edge with `jest.mock`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

<a href="https://app.devin.ai/review/skypilot-org/skypilot/pull/10459" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

### CODE REVIEW

File: sky/dashboard/src/components/jobs.jsx [hunk 1]
```diff
@@ -453,24 +439,21 @@ export function ManagedJobs() {
...
           />
         </div>
       </div>
 
-      <Filters
-        filters={filters}
-        setFilters={setFilters}
-        updateURLParams={updateURLParams}
-      />
+      <Filters filters={filters} setFilters={setFilters} />
```
[the remark this answers is not in the document]
Comment (zpoint) [reply sky/dashboard/src/components/jobs.jsx:476]: Confirmed and fixed in 2bc520b — thanks, this is a real crash and my own verification missed it (I exercised the status pills, the segments and the URL round-trip, but never the filter dropdown itself).

Reproduced on a build of the previous commit: typing in the Name filter and pressing Enter throws `TypeError: u is not a function` and the page goes white. Same for removing a chip and for "Clear filters".

The prop is optional by design — several pages mount these widgets without any URL writer — so the fix makes the callback optional at all four call sites rather than forcing every caller to pass one:

```js
updateURLParams?.(updatedFilters);
```

Two things came out of it beyond the crash:

- `FilterDropdown` hardcoded `[...prevFilters, {property, operator: ':', value: option}]`, which stacks a second chip for single-valued properties. The URL already replaces in that case, so the chip bar and the address bar disagreed. Pages can now pass an `addFilter` reducer, and the jobs page passes the same one it uses for the URL.
- The chip paths had no rendered-interaction coverage at all, which is why CI stayed green. `sky/dashboard/src/components/shared/FilterDropdown.test.jsx` now drives add / remove / clear through the real components; all five fail with `TypeError: updateURLParams is not a function` against the previous code.

File: sky/dashboard/src/components/jobs.jsx [hunk 2]
```diff
@@ -301,7 +319,11 @@ export function ManagedJobs() {
   const jobsRefreshRef = React.useRef(null);
   const poolsRefreshRef = React.useRef(null);
   const [poolsData, setPoolsData] = useState([]);
-  const [filters, setFilters] = useState([]);
+  // Filters and the shareable view state both live in the URL, keyed by name.
+  const { filters, setFilters, view, setView } = useUrlFilterState(
+    JOB_FILTER_SCHEMA,
+    JOB_VIEW_SCHEMA
+  );
```
[the remark this answers is not in the document]
Comment (zpoint) [reply sky/dashboard/src/components/jobs.jsx:353]: Confirmed and fixed in 2bc520b.

Reproduced with a same-route navigation, which is exactly what the pools table's link does:

```js
window.next.router.push('/jobs?property=user&operator=%3A&value=alice')
```

Before: the address bar changed, the chip bar stayed empty and the table kept its previous 10 rows.

The hook only read the query string on mount. It now re-reads on `asPath` changes and skips the writes it made itself, so it doesn't fight its own mirror effect:

```js
const lastWritten = useRef(null);
// ...
if (window.location.search === lastWritten.current) return;
```

Verified after the fix, including that the legacy triple form is normalised on arrival:

| step | URL after | chips | rows |
|---|---|---|---|
| initial | `?owner=all` | — | 10 |
| push `?property=user&operator=:&value=alice` | `?user=alice` | `User : alice` | 9 |
| push `?user=bob` | `?user=bob` | `User : bob` | 6 |

File: sky/dashboard/src/components/jobs.jsx [hunk 3]
```diff
@@ -576,20 +561,40 @@ export function ManagedJobsTable({
...
   const [statusCounts, setStatusCounts] = useState({});
   const [moreMenuOpen, setMoreMenuOpen] = useState(false);
   const moreMenuRef = useRef(null);
   const [controllerStopped, setControllerStopped] = useState(false);
   const [controllerLaunching, setControllerLaunching] = useState(false);
   const [isRestarting, setIsRestarting] = useState(false);
-  const [activeTab, setActiveTab] = useState('all');
-  const [showAllMode, setShowAllMode] = useState(true);
+  // Neither segment is highlighted while specific pills narrow the view.
+  const activeTab = statusGroupName || (statusParam ? null : 'all');
```
[the remark this answers is not in the document]
Comment (zpoint) [reply sky/dashboard/src/components/jobs.jsx]: Confirmed and fixed in 90883f7. Your reasoning is exactly right, and the symptom is a hard failure, not a blank table — `http://.../dashboard/jobs?status=,` renders `Application error: a client-side exception has occurred` with nothing but the nav bar (body text down to 161 characters, zero table rows).

I took your suggestion but pushed it one step further, because the sibling comment about unvalidated values has the same root cause: any `status` the page cannot interpret should simply not narrow the view. So unknown values are dropped first, and the tab then falls back to `all` whenever nothing actually narrows:

```js
const selectedStatuses =
  statusGroupName || !value
    ? []
    : value.split(',').filter((s) => KNOWN_STATUSES.has(s));
// Neither segment is highlighted while specific pills narrow the view.
activeTab: statusGroupName || (selectedStatuses.length > 0 ? null : 'all'),
```

That mirrors how the filter decoders already drop unknown properties, so the two params degrade the same way.

The derivation moved into an exported pure `deriveStatusView(statusParam)` so this is testable without mounting the page. `sky/dashboard/src/components/jobs.statusview.test.jsx` covers `''`, group names, comma lists, `RUNNINGG`, `,`, `,,`, `' '` and `ACTIVE`, and asserts the invariant directly — `activeTab` is either `all` or a key that exists in `statusGroups`. Two of the five fail against the previous derivation.

Verified in a browser after the fix: `?status=,` renders the table with the All segment highlighted, no crash.

File: sky/dashboard/src/components/jobs.jsx [hunk 4]
```diff
@@ -1391,31 +1379,13 @@ export function ManagedJobsTable({
...
-      // We're not in "show all" mode if there are specific statuses selected
-      setShowAllMode(false);
-    }
-
-    // Reset to first page when changing status filters
+    const next = selectedStatuses.includes(status)
+      ? selectedStatuses.filter((s) => s !== status)
+      : [...selectedStatuses, status];
+    setSelectedStatuses(next);
     setCurrentPage(1);
```
[the remark this answers is not in the document]
Comment (zpoint) [reply sky/dashboard/src/components/jobs.jsx:1412]: You're right that the round trip changes, and it's worth being precise about when. On master the deselect path is:

```js
if (newSelectedStatuses.length === 0) {
  // When deselecting the last selected status, go back to "show all" mode
  // for the current active tab (active/finished)
  setShowAllMode(true);
  setSelectedStatuses([]);
}
```

so it returns to whatever `activeTab` happened to be. Since `activeTab` defaults to `'all'` (`useState('all')`), the common flow — arrive, click RUNNING, click RUNNING again — lands on every status on master too. The divergence is only when the user first clicked the **Active** segment: master returns to Active, this PR widens to every status.

I'm keeping the widening, deliberately:

- Master's behaviour depends on `activeTab` + `showAllMode`, two pieces of hidden component state. Reproducing it means the URL is no longer sufficient to describe the view, which is the thing this PR is removing. A pill's group can't recover it either — `RUNNING ∈ active`, so deriving the group from the pill would return Active even for a user who came from All, i.e. a silent *narrowing*, which is the worse direction to guess wrong in.
- It isn't silent in the UI: `activeTab` becomes `'all'`, so the **All** segment lights up at the same moment the finished jobs appear.

Documented in the PR description under the pill/segment unification, since that was the actual gap you identified. If it reads badly in practice, the honest fix is a URL that carries both the group and the pills, and that's a follow-up rather than reinstated hidden state.

File: sky/dashboard/src/components/jobs.jsx [hunk 5]
```diff
@@ -576,20 +561,40 @@ export function ManagedJobsTable({
...
+  const statusGroupName = STATUS_GROUP_NAMES.includes(statusParam)
+    ? statusParam
+    : null;
+  const selectedStatuses = React.useMemo(
+    () =>
+      statusGroupName || !statusParam
+        ? []
+        : statusParam.split(',').filter(Boolean),
+    [statusParam, statusGroupName]
+  );
```
[the remark this answers is not in the document]
Comment (zpoint) [reply sky/dashboard/src/components/jobs.jsx]: Confirmed and fixed in 90883f7 — and it's slightly worse than "sent verbatim". Checked live with `?status=RUNNINGG`:

- the value renders as a **status chip in the pill bar** with a count of 0, next to the real statuses, because `nonPrimarySelectedStatuses` is `selectedStatuses` minus the primary list and never checks membership;
- the server accepts it and returns nothing, so the table shows "No active jobs" — no error, just a confusing empty view with a fake status on screen.

Fixed by filtering the split against the known statuses, which is the same fix as the `?status=,` crash you flagged:

```js
const KNOWN_STATUSES = new Set([...PRIMARY_STATUSES, ...OTHER_STATUSES]);
// ...
: value.split(',').filter((s) => KNOWN_STATUSES.has(s));
```

(`PRIMARY_STATUSES ∪ OTHER_STATUSES` is the same 13-element set as `statusGroups.active ∪ statusGroups.finished`, so nothing legitimate is dropped.)

A bad link now degrades the way an unknown filter property already does — ignored, All highlighted, table populated. Covered in `jobs.statusview.test.jsx`, and verified in a browser: the fake chip is gone and the table renders.

**Comment by zpoint** [comment]
Thanks for the review — all five went through a live check against a local server with seeded jobs, not just a read of the diff. Three were real bugs I introduced and are fixed; two were judgement calls, one of which turned out to also be a bug. Per-comment detail is inline; summary here.

| # | Issue | Verdict | Fix |
|---|---|---|---|
| 1 | Adding/removing a filter throws | **Real crash** | `2bc520b` |
| 2 | "See all jobs" link no longer filters | **Real** | `2bc520b` |
| 3 | `?status=,` fails to render | **Real crash** | `90883f7` |
| 4 | Deselecting the last pill widens to every status | **Deliberate**, now documented | — |
| 5 | Unvalidated status values forwarded to the server | **Real**, worse than described | `90883f7` |

**#1** is the one I'm least happy about. My verification covered the status pills, the segments and the URL round-trip, but never the filter dropdown itself, so I shipped a page that white-screened on every chip interaction. Reproduced before fixing: `TypeError: u is not a function`. The prop is optional by design, so the call sites are now optional rather than the prop being made mandatory.

**#3** is a hard failure, not a blank table: `?status=,` renders `Application error: a client-side exception has occurred` with nothing but the nav bar.

**#5** is worse than "sent verbatim" — `?status=RUNNINGG` also draws a fake zero-count status chip into the pill bar. Same root cause as #3, so both are fixed by ignoring status values the page cannot interpret, which is how the filter decoders already treat unknown properties.

**#4** I'm keeping. On master, deselecting the last pill restores `activeTab`, which defaults to `'all'` — so the common flow is unchanged, and the divergence only affects a user who first clicked **Active**. Reproducing master exactly means keeping the group and the pills in separate hidden state, which is what this PR removes; and the change is visible, since the **All** segment lights up as the finished jobs appear. Written into the PR description.

On coverage: the reason CI stayed green through #1 is that the `dashboard` job runs `lint`, `format:check` and `build` and never invokes jest, so no test renders these components. Two suites now cover the gaps this review found, and both fail against the previous code:

- `shared/FilterDropdown.test.jsx` — chip add / remove / clear through the real components (5 tests, all 5 fail before the fix).
- `jobs.statusview.test.jsx` — the `status` values a hand-edited or stale link can carry, asserting that the activity tab is always either `all` or a key that exists in `statusGroups` (2 of 5 fail before the fix).

Wiring jest into that job is worth doing but is not this PR: two suites already fail on master, and one of them (`data/connectors/jobs.test.jsx`) is a module cycle between `data/connectors/jobs` and the plugin data-enhancement module, which is also why the new jobs test has to cut that edge with `jest.mock`. Left as follow-up rather than smuggled in here.

**Review by DanielZhangQD — APPROVED** [review]
LGTM! Thanks! @zpoint
